activejob/lib/active_job/serializers.rb

```diff
@@ -55,16 +56,26 @@ def serializers
       def add_serializers(*new_serializers)
         self._additional_serializers += new_serializers.flatten
       end
+
+      # Adds a new serializer that will be used to serialize a specific class. These serializers do not consider
+      # ancestors, and will only be used if the class matches exactly. Class-based serializer lookups are much faster
+      # than regular serializer lookups, which have to iterate through all serializers to find one that can handle the
+      # given object. Most regular serializers call the object's `#is_a?` method to determine if they can handle the
```

Comment:
```suggestion
      # given object. Most regular serializers call the object's `#is_a?` method to determine if they can handle the
```

The performance of `is_a?` is rather negligible if the argument is a class. It used to be quite costly , but was optimized a few years ago in https://github.com/ruby/ruby/pull/5568

Reply:
> I wonder then if the majority of the perf improvement is just removing a layer of method dispatch?

Not quite: https://share.firefox.dev/3USDwD6

The current implementation is a bit silly indeed because it's basically:

```ruby
serializers.each do |s|
  return s.serialize(obj) if s.serialize?(obj)
end
```

In your specific benchmark, 66% of the serialization time is just calling `serialize?`, which is agravated by the fast that the default `serialize?` implementation call another method.

e.g.
```ruby
def klass
  Time
end

def serialize?(argument)
  argument.is_a?(klass)
end

serialize(obj)
```

Which is way slower than directly calling `obj.is_a?(Time)`, YJIT can optimize the shit out of the later.

But you patch also help by turning a largely `O(n)` algo into a mostly `O(1)` one.

However I don't think we need a new API to implement that, we can call `#klass` when a new serializer is registered and then have the lookup be:

```ruby
serializer = serializers_by_klass[obj.class] || serializers.detect { ... }
```

But also some types like Symbol should just be "builtin", no need to go through a dispatch for them, they should be handled in the main loop.

Then, there are some serializers, like TimeWithZone that are inherently slow and can be sped up 10x: https://github.com/rails/rails/compare/rails:1dda631...byroot:1e9cf20

And finally, the serialization format isn't ideal, so I'm quite tempted to introduce format versioning and introduce a new more efficient format. Since we're meeting with the core team next week, I'll likely mention it there.